### PR TITLE
🧷  Making the entire input shape fixed

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes
1, Implement the [make-dynamic-shape-fixed](https://onnxruntime.ai/docs/tutorials/mobile/helpers/make-dynamic-shape-fixed.html) in Olive.
2. Update related tests and docs.
3. Add e2e example for bert-snpe also fix the label type judgement in qnn/snpe evaluator.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/passes/onnx/dynamic_to_fixed_shape.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+from typing import Any, Callable, Dict, List
+
+from onnxruntime.tools.onnx_model_utils import fix_output_shapes, make_dim_param_fixed, make_input_shape_fixed
```
Comment (guotuofeng) [inline olive/passes/onnx/dynamic_to_fixed_shape.py:9]: move the import to the run config to decouple the dependency of this pass with ort. From the time being, the setup could be run without ORT installed.

File: docs/source/features/passes/onnx.md [hunk 2]
```diff
@@ -411,3 +411,39 @@ b. More fine-grained control of the conversion conditions is also possible:
...
+For example, often models have a dynamic batch size so that training is more efficient. In mobile scenarios the batch generally has a size of 1. Making the batch size dimension ‘fixed’ by setting it to 1 may allow NNAPI and CoreML to run of the model.
+
+The helper can be used to update specific dimensions, or the entire input shape.
+
+### Example Configuration
+
+a. Making a symbolic dimension fixed
+```json
+{
+    "type": "OrtMixedPrecision",
```
Comment (jambayk) [inline docs/source/features/passes/onnx.md:428]: Pass name is incrorrect

File: examples/bert/bert_ptq_qnn_ep.json [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            "type": "OnnxStaticQuantization",
+            "config": {
+                "prepare_qnn_config": true,
+                "activation_type": "QUInt16",
+                "weight_type": "QUInt8",
+                "calibrate_method": "MinMax",
+                "data_config": "__input_model_data_config__"
+            }
+        },
+        "perf_tuning": {
```
Comment (jambayk) [inline examples/bert/bert_ptq_qnn_ep.json:72]: Since qnn ep cannot be installed in the same environment as Olive, the perf tuning pass will not work. Did you test this workflow?

Comment (jambayk) [reply examples/bert/bert_ptq_qnn_ep.json:72]: Also, I am not sure if qnn ep can be tuned the same way as other ep.

File: examples/bert/bert_ptq_qnn_ep.json [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    "subset": "mrpc",
+                    "split": "validation",
+                    "input_cols": ["sentence1", "sentence2"],
+                    "label_cols": ["label"],
+                    "batch_size": 1
+                }
+            }
+        }
+    },
+    "evaluators": {
```
Comment (jambayk) [inline examples/bert/bert_ptq_qnn_ep.json:19]: Same for the evaluations. I am currently working on updating the support for qnn ep evaluations. 

I think it might be better to just remove this example for now and add it when the e2e workflows has been validated for qnn ep. Especially, since this example is not related to the PR.

File: examples/bert/bert_ptq_qnn_ep.json (hunk 3 above)
Comment (trajepl) [reply examples/bert/bert_ptq_qnn_ep.json:72]: Sorry, this config file is added from my local test on cpu EP. Will remove it.

**Review by jambayk — APPROVED** [review]

File: olive/passes/onnx/dynamic_to_fixed_shape.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                ),
+            ),
+        }
+        config.update(get_external_data_config())
+        return config
+
+    @staticmethod
+    def _validators() -> Dict[str, Callable[..., Any]]:
+        return {
+            "validate_dim_param_and_value": root_validator(allow_reuse=True)(_validate_dim_param_and_value),
```
Comment (guotuofeng) [inline olive/passes/onnx/dynamic_to_fixed_shape.py:60]: for root_validator, could we use only one?

Comment (trajepl) [reply olive/passes/onnx/dynamic_to_fixed_shape.py]: Updated. Also rebase from main branch to resolve the docs building error.

**Review by jambayk — APPROVED** [review]

**Comment by guotuofeng** [comment]
are there any usage example? I don't see it is used in any examples.

**Review by guotuofeng — CHANGES_REQUESTED** [review]
Please help update the examples to use this pass.

**Comment by trajepl** [comment]
> Please help update the examples to use this pass.

Add e2e example for bert-snpe also fix the label type judgement in qnn/snpe evaluator.

File: olive/passes/onnx/dynamic_to_fixed_shape.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    if (not values["input_name"]) ^ (not values["input_shape"]):
+        raise ValueError("input_name and input_shape must be both provided or both None.")
+
+    if values["dim_param"] and values["dim_value"]:
+        if len(values["dim_param"]) != len(values["dim_value"]):
+            raise ValueError("dim_param and dim_value must have the same number of elements.")
+        if any(i <= 0 for i in values["dim_value"]):
+            raise ValueError("dim_value must be all > 0 when dim_param is provided.")
+
+    if values["input_name"] and values["input_shape"]:
```
Comment (xiaoyu-work) [inline olive/passes/onnx/dynamic_to_fixed_shape.py:105]: Nit: can be abstracted to avoid duplicated code.

Comment (trajepl) [reply olive/passes/onnx/dynamic_to_fixed_shape.py:105]: Thanks for your suggestion~ 

We only make the validation for two-combinations(input_name&input_shape, dim_param&dim_value) here. So will just keep the original logics. If there is further similar logics, I will try to extract the common part.

**Review by guotuofeng — APPROVED** [review]
